File: leetcode/Medium/rotate-array/rotate-array.cpp
class Solution {
public:
void reverse(int i, int j, vector<int>&a){
        while(i<=j){
            int temp = a[i];
            a[i] = a[j];
            a[j] = temp;
            i++;
            j--;
        } 
        return;
    }
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k = k%n;
        reverse(0 , n-k-1, nums);
        reverse(n-k, n-1, nums);
        reverse(0 , n-1, nums);
    }
};

File: leetcode/Medium/rotate-array/rotate-array.py
class Solution(object):
    def rotate(self, nums, k):
        n = len(nums)
        if n == 0:
            return 
        k %= n
        if k == 0:
            return

        # m-1(Brute force)
        # for i in range(1, k+1):
        #     temp = nums[n-1]
        #     for j in range(n-2,-1,-1):
        #         nums[j+1] = nums[j]
        #     nums[0] = temp

        # m-2(Optimal solution)
        def reverse(nums, l, r):
            if r <= l:
                return
            i = l
            j = r
            while i < j:
                nums[i], nums[j] = nums[j], nums[i]
                j -= 1
                i += 1
            
        reverse(nums, 0, n-1)
        reverse(nums, 0, k-1)
        reverse(nums, k, n-1)